doc/man/man8/slconfd.8.man
'\" rtp
.\" vim: ft=nroff sw=4 noet nocin nosi com=b\:.\\\" fo+=tcqlorn tw=77
.\" =========================================================================
.\"
.\" @(#) doc/man/man8/slconfd.8.man
.\"
.\" =========================================================================
.\"
.\" Copyright (c) 2008-2015  Monavacon Limited <http://www.monavacon.com/>
.\" Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
.\" Copyright (c) 1997-2001  Brian F. G. Bidulock <bidulock@openss7.org>
.\"
.\" All Rights Reserved.
.\"
.\" Permission is granted to copy, distribute and/or modify this manual under
.\" the terms of the GNU Free Documentation License, Version 1.3 or any later
.\" version published by the Free Software Foundation; with no Invariant
.\" Sections, no Front-Cover Texts, and no Back-Cover Texts.  A copy of the
.\" license is included in the section entitled "GNU Free Documentation
.\" License".
.\"
.\" Permission to use, copy and distribute this manual without modification,
.\" for any purpose and without fee or royalty is hereby granted, provided
.\" that both the above copyright notice and this permission notice appears
.\" in all copies and that the name of OpenSS7 Corporation not be used in
.\" advertising or publicity pertaining to distribution of this documentation
.\" or its contents without specific, written prior permission.  OpenSS7
.\" Corporation makes no representation about the suitability of this manual
.\" for any purpose.  It is provided "as is" without express or implied
.\" warranty.
.\"
.\" Permission is granted to process this file through groff and print the
.\" results, provided the printed document carries a copying permission
.\" notice identical to this one except for the removal of this paragraph
.\" (this paragraph not being relevant to the printed manual).
.\"
.\" OPENSS7 CORPORATION DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS MANUAL
.\" INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
.\" PARTICULAR PURPOSE, NON-INFRINGEMENT, OR TITLE; THAT THE CONTENTS OF THE
.\" DOCUMENT ARE SUITABLE FOR ANY PURPOSE, OR THAT THE IMPLEMENTATION OF SUCH
.\" CONTENTS WILL NOT INFRINGE ON ANY THIRD PARTY PATENTS, COPYRIGHTS,
.\" TRADEMARKS OR OTHER RIGHTS.  IN NO EVENT SHALL OPENSS7 CORPORATION BE
.\" LIABLE FOR ANY DIRECT, INDIRECT, SPECIAL OR CONSEQUENTIAL DAMAGES OR ANY
.\" DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER
.\" IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING
.\" OUT OF OR IN CONNECTION WITH ANY USE OF THIS DOCUMENT OR THE PERFORMANCE
.\" OR IMPLEMENTATION OF THE CONTENTS THEREOF.
.\" 
.\" Since the Linux kernel and libraries are constantly changing, this manual
.\" page may be incorrect or out-of-date.  The author(s) assume no
.\" responsibility for errors or omissions, or for damages resulting from the
.\" use of the information contained herein.  The author(s) may not have
.\" taken the same level of care in the production of this manual, which is
.\" licensed free of charge, as they might when working professionally.  The
.\" author(s) will take no responsibility in it.
.\" 
.\" Formatted or processed versions of this manual, if unaccompanied by the
.\" source, must acknowledge the copyright and authors of this work.
.\"
.\" -------------------------------------------------------------------------
.\"
.\" U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
.\" behalf of the U.S. Government ("Government"), the following provisions
.\" apply to you.  If the Software is supplied by the Department of Defense
.\" ("DoD"), it is classified as "Commercial Computer Software" under
.\" paragraph 252.227-7014 of the DoD Supplement to the Federal Acquisition
.\" Regulations ("DFARS") (or any successor regulations) and the Government
.\" is acquiring only the license rights granted herein (the license rights
.\" customarily provided to non-Government users).  If the Software is
.\" supplied to any unit or agency of the Government other than DoD, it is
.\" classified as "Restricted Computer Software" and the Government's rights
.\" in the Software are defined in paragraph 52.227-19 of the Federal
.\" Acquisition Regulations ("FAR") (or any successor regulations) or, in the
.\" cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the
.\" FAR (or any successor regulations).
.\"
.\" =========================================================================
.\" 
.\" Commercial licensing and support of this software is available from
.\" OpenSS7 Corporation at a fee.  See http://www.openss7.com/
.\" 
.\" =========================================================================
.\"
.R1
bracket-label "\fR[\fB" "\fR]" "\fR, \fB"
no-default-database
database openss7.refs
accumulate
move-punctuation
abbreviate A
join-authors ", " ", " " and "
et-al " et al" 2 3
abbreviate-label-ranges ".."
sort-adjacent-labels
search-truncate 40
search-ignore CGIQOSTU
.R2
.so openss7.macros
.\"
.\"
.TH SLCONFD 8 "@PACKAGE_DATE@" "@PACKAGE@-@VERSION@" "@PACKAGE_TITLE@ Administration"
.SH NAME
.B slconfd
\- Signalling Link Configuration Daemon
.\"
.\"
.SH SYNOPSIS
.PP
.HP 12
\fBslconfd\fR [\fIoptions\fR]
.PD 0
.HP
\fBslconfd\fR {\fB-h\fR|\fB--help\fR}
.HP
\fBslconfd\fR {\fB-V\fR|\fB--version\fR}
.HP
\fBslconfd\fR {\fB-C\fR|\fB--copying\fR}
.PD
.\"
.\"
.SH DESCRIPTION
.PP
.B slconfd
is a configuration daemon for the SL-MUX multiplexing driver,
.BR sl-mux (4).
Its purpose is to open Signalling Link streams on various drivers, attach them
to specific PPAs, and link the SL streams under the SL-MUX multiplexing
driver, configure each SL lower multiplex Stream, and then wait in the
background listening for management events on a management stream.  Also, the
daemon watches for updates to the configuration file and applies the updates
as they occur.
.PP
The
.B slconfd
is intended to be run at system startup from a System V init script.  Once the
configuration of the SL-MUX multiplexing driver is complete, applications
programs can open, attach, enable and use Signalling Link streams opened on
the driver.
.PP
The purpose of the SL-MUX driver is to provide a unified interface for
accessing signalling links of a number of forms and from a number of various
drivers.  Signalling links could correspond to device drivers, or could be
M2UA signalling links or other implementations.  M2PA signalling links require
some additional configuration before they can be used and will be linked under
the SL-MUX by a separate M2PA configuration daemon.
.\"
.\"
.SH OPTIONS
.PP
.TP
.BR -h ", " --help
When this option is encountered, print usage information to
.I stdout
and exit.
.TP
.BR -V ", " --version
When this option is encountered, print version information to
.I stdout
and exit.
.TP
.BR -C ", " --copying
When this option is encountered, print copyright and copying information to
.I stdout
and exit.
.\"
.\"
.SH "FILE FORMAT"
.PP
When the
.BR -f " or " --file
.I filename
option is specified, the supplied file contains SL configuration entries as
follows:
.IP \(bu \w'\(bu\(em'u
Each line in the file contains one entry or comment per line.
.IP \(bu
Each entry consists of a number of fields delimited by whitespace (see
.BR isspace (3)).
.IP \(bu
Any field that begins with an octothorpe,
.RI ` # ',
character is a comment that continues to the end of the line.
.IP \(bu
Any line containing only a comment (beginning with optional whitespace
followed by an octothorpe,
.RI ` # ',
character) is ignored.
.PP
Each line containing an SL configuration specification entry consists of the
following fields:
.TP
(1) \fIclei\fR
A common language equipment identifier for the signalling link.
.TP
(2) \fIglobal-ppa\fR
A global Physical Point of Appearance.  This PPA can be used by streams
opening the SL-MUX to attach to this signalling link.
.TP
(3) \fIdevice-type\fR
The device type defines the type of device that is providing the signalling
link.  The device type can currently be one of:
.BR x400 ", " m2pa " or " m2ua .
The remaining fields are interpreted depending on the device type as follows:
.RS
.PP
When field (3) is
.BR x400 ,
the remaining fields are interpreted as follows:
.TP
(3) \fBx400\fR
.TP
(4) \fIdevice\fR
The device driver to open to obtain a serving signalling link.
This is the filesystem path to the appropriate device node.
.TP
(5) \fIlocal-ppa\fR
If the device above is a Style II device, the local Physical Point of
Attachment to which to attach the signalling link before it is linked under
the SL-MUX multiplexing driver.  If the device above is a Style I device, this
field should be a single dash,
.RI ` - '.
.PP
When field (3) is
.BR m2ua ,
the remaining fields are interpreted as follows:
.TP
(3) \fBm2ua\fR
.TP
(4) \fIdevice\fR
The device driver to open to obtain a serving signalling link.  This is the
filesystem path to the appropriate device node.
.TP
(5) \fIlocal-ppa\fR
The local Physical Point of Attachment to which to attach the signalling link
before it is linked under the SL-MUX multiplexing driver.  An m2ua device is
always a Style II device.
.PP
When field (3) is
.BR m2pa ,
the remaining fields are interpreted as follows:
.TP
(3) \fBm2pa\fR
.TP
(4) \fIdevice\fR
The device driver to open to obtain a serving transport provider.
This is the filesystem path to the appropriate device node.
.TP
(5) \fImodules\fR
A comma separated list of modules to pushed over the transport provider to
effect a signalling link.
.TP
(6) \fIlocal-address\fR
The local address to which to bind the transport provider.  This address can
be a hostname and service name or a numbers and dots notation IP address and
port number.
.TP
(7) \fIremote-address\fR
The remote address to which to connect the trasnport provider.  This address
can be a hostname and service name or a numbers and dots notation IP address
and port number.
.RE
.PP
.B slconfd
opens the configuration file and reads the entries line by line.  Each entry
is indepenendent of other entries.  For each entry,
.B slconfd
performs the following actions:
.IP 1. \w'0.\(em'u
The daemon opens the device specified in the
.I device
field.
.IP 2.
If the
.I local-ppa
field is other than a signle dash, the daemon inteprets the field as a integer
number and uses the integer number in an
.BR LMI_ATTACH_REQ (7)
request to attach the opened Stream to the specific signalling link.
.IP 3.
The daemon links the signalling link under the SL-MUX multiplexing drivers
using the
.BR I_PLINK (7)
input-output control.
.IP 4.
The daemon uses SL-MUX specific input-output controls to assign the
.I global-ppa
and
.I clei
to the linked singalling link.  Streams opened on the SL-MUX driver can then
attach to this signalling link using the global PPA or the CLEI for the
signalling link.
.PP
Once the
.B slconfd
has linked and configured all of the signalling link streams specified in the
configuration file, it listens on the SL-MUX management stream for management
information, taking appropriate actions.
The daemon also watches the configuration file for changes and effects those
changes when they occur.  Changes are effected by linking additional
signalling links or unlinking existing signalling links to alter the
configuration of the SL-MUX to match that of the altered configuration file.
.\"
.\"
.SH SIGNALS
.PP
During its operation the
.B slconfd
daemon accepts a number of signals as follows:
.\"
.\"
.SH DIAGNOSTICS
.PP
An exit status of zero (0) indicates that the command was successful; one (1)
indicates that an error occurred and a diagnostic message is printed to
.IR stderr ;
two (2) indicates that the option syntax was in error and a diagnostic message
is printed on
.IR stderr .
.PP
The
.B --quiet
option suppresses the printing of diagnostic messages to
.IR stderr .
.\"
.\"
.SH NOTICES
.PP
Other programs may open and parse the configuration file to determine the
global PPA associated with a given Common Lanuage Equipment Identifier.  The
SL-MUX also supports a specialized input-output control that enables an
application program to provide a Common Lanuage Equiment Identifier and have
the driver translate that string to a global PPA.  The SL-MUX also supports
using the CLEI as the PPA in a
.BR LMI_ATTACH_REQ (7)
primitive.
.\"
.\"
.\".SH EXAMPLES
.\".PP
.\"
.\"
.\".SH DEVICES
.\".PP
.\"
.\"
.\".SH FILES
.\".PP
.\"
.\"
.SH "SEE ALSO"
.PP
.BR sl-mux (4).
.\"
.\"
.SH BUGS
.PP
.B slconfd
has no known bugs.
.\"
.\"
.SH COMPATIBILITY
.PP
.B slconfd
is an OpenSS7-specific configuration command.  Because this command is used at
startup and is transparent to the SL-User opening SL streams, compatibility is
not an issue.
.\"
.\"
.SH CONFORMANCE
.PP
System configuration programs are not subject to standardization.
.B slconfd
conforms to OpenSS7 specifications.
.\"
.\"
.SH HISTORY
.PP
.B slconfd
first appeared in the OpenSS7
.B strss7-0.9a.6
package.
.\"
.\"
.[
$LIST$
.]
.TI



NEWS
OpenSS7 -- history of user visible changes.  2015-01-15
$Id: NEWS,v 1.1.2.25 2011-05-10 13:45:30 brian Exp $
Copyright (c) 2008-2015  Monavacon Limited. <http://www.monavacon.com/>
Copyright (c) 2001-2008  OpenSS7 Corporation. <http://www.openss7.com/>
Copyright (c) 1997-2001  Brian Bidulock <bidulock@openss7.org>
See the end for copying conditions (for this file).

The sections that follow provide information on OpenSS7 releases of the
'OpenSS7' package.

Major changes for release openss7-1.1.7.20150115
------------------------------------------------

This is the seventh release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Source code release on GitHub.

Major changes for release openss7-1.1.7.20131209
------------------------------------------------

This is the seventh release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Build corrections for RHEL 6.4.

   - Scripts for country boundary, shoreline and rivers database
     preparation.

   - Database for ANSI T1.101 point code assignments.

   - Suppress extransous /proc/1/comm error messages when testing for
     systemd.

   - Handle PUIAS to Springdale rename for PUIAS distribution.

Major changes for release openss7-1.1.7.20131123
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Correction to ca-cert handling in RHEL 6 install scriptlet.

   - RHEL 6 build changes.

   - Set permissions correctly when applying kernel module patch.

   - Do not duplicate certs when including ca-cert.

   - Avoid conflicting shelll variable in RPM install scriplet.

   - Remove /proc/1/comm checks for systemd in module loader.

Major changes for release openss7-1.1.7.20130209
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Corrections to RPM spec file.

Major changes for release openss7-1.1.7.20130129
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Compensate for RHEL6 lack of c_rehash.

   - Do not create sysfs nodes: they do not work properly.

   - Use no foreground flag in strace.service file.

   - More error traces for strerr logger.

   - Remove static device names (the do not work either).

   - Changes to support weak kernel modules.

   - Clean up weak module support.

Major changes for release openss7-1.1.7.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Better Archlinux install script support.

Major changes for release openss7-1.1.6.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Remove use of kill_litter_super().

Major changes for release openss7-1.1.3.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Updates and corrections to weak kernel module builds.

   - Repo support and corrections for Archlinux packages.

Major changes for release openss7-1.1.3.20130123
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Additional support for Archlinux.

   - Add udev rules for strace and strerr.

   - Add system unit files for strace and strerr.

   - Add repo support for Archlinux packages.

   - Build Archlinux packages better.

Major changes for release openss7-1.1.3.20130121
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Build support for SLES 11.

   - Handle certificates better.

   - Get configuration files in the right places for SLES 11.

   - Support in RPM spec file for certificates on SLES 11.

   - Create a new OpenSS7 certificate.

   - Repository support for SLES 11.

   - RPM spec file reworked for new RPM release.

   - Better certificate handling for RHEL.

   - Additional build support for RHEL 6 with certificates.


src/util/slconfd.c
/*****************************************************************************

 @(#) File: src/util/slconfd.c

 -----------------------------------------------------------------------------

 Copyright (c) 2008-2015  Monavacon Limited <http://www.monavacon.com/>
 Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
 Copyright (c) 1997-2001  Brian F. G. Bidulock <bidulock@openss7.org>

 All Rights Reserved.

 This program is free software: you can redistribute it and/or modify it under
 the terms of the GNU Affero General Public License as published by the Free
 Software Foundation, version 3 of the license.

 This program is distributed in the hope that it will be useful, but WITHOUT
 ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 FOR A PARTICULAR PURPOSE.  See the GNU Affero General Public License for more
 details.

 You should have received a copy of the GNU Affero General Public License
 along with this program.  If not, see <http://www.gnu.org/licenses/>, or
 write to the Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA
 02139, USA.

 -----------------------------------------------------------------------------

 U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
 behalf of the U.S. Government ("Government"), the following provisions apply
 to you.  If the Software is supplied by the Department of Defense ("DoD"), it
 is classified as "Commercial Computer Software" under paragraph 252.227-7014
 of the DoD Supplement to the Federal Acquisition Regulations ("DFARS") (or any
 successor regulations) and the Government is acquiring only the license rights
 granted herein (the license rights customarily provided to non-Government
 users).  If the Software is supplied to any unit or agency of the Government
 other than DoD, it is classified as "Restricted Computer Software" and the
 Government's rights in the Software are defined in paragraph 52.227-19 of the
 Federal Acquisition Regulations ("FAR") (or any successor regulations) or, in
 the cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the FAR
 (or any successor regulations).

 -----------------------------------------------------------------------------

 Commercial licensing and support of this software is available from OpenSS7
 Corporation at a fee.  See http://www.openss7.com/

 *****************************************************************************/

static char const ident[] = "src/util/slconfd.c (" PACKAGE_ENVR ") " PACKAGE_DATE;

/*
 *  This is a configuration daemon for the SL-MUX multiplexing driver.  Its purpose is to open
 *  Signalling Link streams on various drivers, attach them to specific PPAs, and link the SL
 *  streams under the SL-MUX multiplexing driver, configure each SL lower multiplex Stream, and then
 *  wait in the background listening for management events on a management stream.  Also, the daemon
 *  watches for updates for the configuration file and applies the updates as they occur.
 *
 *  The daemon is intended to run at system startup from a System V init script.  Once the
 *  configuration of the SL-MUX multiplexing driver is complete, applications programs can open,
 *  attach, enable and use Signalling Link streams opened on the driver.
 *
 *  The purpose of the SL-MUX driver is to provide a unified interface for accessing signalling
 *  links of a number of forms and from a number of various drivers.  Signalling links could
 *  correspond to device drivers, or could be M2UA or M2PA signalling links, or other
 *  implementations.  M2PA signalling links require some additional configuration before they can be
 *  used and will be linked under the SL-MUX by a separate M2PA configuration daemon.
 *
 *  Another possibility is to implement plugable dlopened modules for various devices.  A plugable
 *  module would define its own syntax for the last fields in the configuration file entry (after
 *  the selector).  New selectors could then be defined.
 */

#include <stropts.h>
#include <stdlib.h>
#include <unistd.h>
#include <stdio.h>
#include <string.h>
#include <errno.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/poll.h>
#include <sys/time.h>
#include <sys/ioctl.h>
#include <fcntl.h>
#include <getopt.h>
#include <time.h>
#include <signal.h>
#include <syslog.h>
#include <sys/utsname.h>

#include <ss7/lmi.h>
#include <ss7/lmi_ioctl.h>
#include <ss7/sdli.h>
#include <ss7/sdli_ioctl.h>
#include <ss7/sdti.h>
#include <ss7/sdti_ioctl.h>
#include <ss7/sli.h>
#include <ss7/sli_ioctl.h>

int output = 1;
int debug = 0;


#define BUFSIZE 512

char cbuf[BUFSIZE];
char dbuf[BUFSIZE];

struct strbuf ctrl = { sizeof(cbuf), 0, cbuf };
struct strbuf data = { sizeof(dbuf), 0, dbuf };

struct strioctl ctl;

static void
copying(int argc, char *argv[])
{
	if (!output)
		return;
	(void) fprintf(stdout, "\
--------------------------------------------------------------------------------\n\
%1$s\n\
--------------------------------------------------------------------------------\n\
Copyright (c) 2008-2015  Monavacon Limited <http://www.monavacon.com/>\n\
Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>\n\
Copyright (c) 1997-2001  Brian F. G. Bidulock <bidulock@openss7.org>\n\
\n\
All Rights Reserved.\n\
--------------------------------------------------------------------------------\n\
This program is free software: you can  redistribute it  and/or modify  it under\n\
the terms of the  GNU Affero  General  Public  License  as published by the Free\n\
Software Foundation, version 3 of the license.\n\
\n\
This program is distributed in the hope that it will  be useful, but WITHOUT ANY\n\
WARRANTY; without even  the implied warranty of MERCHANTABILITY or FITNESS FOR A\n\
PARTICULAR PURPOSE.  See the GNU Affero General Public License for more details.\n\
\n\
You should have received a copy of the  GNU Affero General Public License  along\n\
with this program.   If not, see <http://www.gnu.org/licenses/>, or write to the\n\
Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.\n\
--------------------------------------------------------------------------------\n\
U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on behalf\n\
of the U.S. Government (\"Government\"), the following provisions apply to you. If\n\
the Software is supplied by the Department of Defense (\"DoD\"), it is classified\n\
as \"Commercial  Computer  Software\"  under  paragraph  252.227-7014  of the  DoD\n\
Supplement  to the  Federal Acquisition Regulations  (\"DFARS\") (or any successor\n\
regulations) and the  Government  is acquiring  only the  license rights granted\n\
herein (the license rights customarily provided to non-Government users). If the\n\
Software is supplied to any unit or agency of the Government  other than DoD, it\n\
is  classified as  \"Restricted Computer Software\" and the Government's rights in\n\
the Software  are defined  in  paragraph 52.227-19  of the  Federal  Acquisition\n\
Regulations (\"FAR\")  (or any successor regulations) or, in the cases of NASA, in\n\
paragraph  18.52.227-86 of  the  NASA  Supplement  to the FAR (or any  successor\n\
regulations).\n\
--------------------------------------------------------------------------------\n\
Commercial  licensing  and  support of this  software is  available from OpenSS7\n\
Corporation at a fee.  See http://www.openss7.com/\n\
--------------------------------------------------------------------------------\n\
", ident);
}

static void
version(int argc, char *argv[])
{
	if (!output)
		return;
	(void) fprintf(stdout, "\
%1$s (OpenSS7 %2$s) %3$s (%4$s)\n\
Written by Brian Bidulock.\n\
\n\
Copyright (c) 2008, 2009, 2010, 2011, 2015  Monavacon Limited.\n\
Copyright (c) 2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008  OpenSS7 Corporation.\n\
Copyright (c) 1997, 1998, 1999, 2000, 2001  Brian F. G. Bidulock.\n\
This is free software; see the source for copying conditions.  There is NO\n\
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.\n\
\n\
Distributed by OpenSS7 under GNU Affero General Public License Version 3,\n\
with conditions, incorporated herein by reference.\n\
\n\
See `%1$s --copying' for copying permissions.\n\
", NAME, PACKAGE, VERSION, PACKAGE_ENVR " " PACKAGE_DATE);
}

static void
usage(int argc, char *argv[])
{
	if (!output)
		return;
	(void) fprintf(stderr, "\
Usage:\n\
    %1$s [options] {{-p|--ppa} PPA | {-c|--clei} CLEI}\n\
    %1$s {-h|--help}\n\
    %1$s {-V|--version}\n\
    %1$s {-C|--copying}\n\
", argv[0]);
}

static void
help(int argc, char *argv[])
{
	if (!output)
		return;
	(void) fprintf(stdout, "\
Usage:\n\
    %1$s [options] {{-p|--ppa} PPA | {-c|--clei} CLEI}\n\
    %1$s {-h|--help}\n\
    %1$s {-V|--version}\n\
    %1$s {-C|--copying}\n\
Arguments:\n\
    None.\n\
Options:\n\
    -q, --quiet\n\
        suppress normal output (equivalent to --verbose=0)\n\
    -d, --debug [LEVEL]\n\
        increment or set debug LEVEL [default: 0]\n\
    -v, --verbose [LEVEL]\n\
        increment or set output verbosity LEVEL [default: 1]\n\
        this option may be repeated.\n\
    -h, --help, -?, --?\n\
        print this usage information and exit\n\
    -V, --version\n\
        print version and exit\n\
    -C, --copying\n\
        print copying permission and exit\n\
", argv[0]);
}

int
main(int argc, char **argv)
{
	int c;
	int val;

	while (1) {
		int option_index = 0;
		/* *INDENT-OFF* */
		static struct option long_options[] = {
			{"quiet",	no_argument,		NULL, 'q'},
			{"verbose",	optional_argument,	NULL, 'v'},
			{"help",	no_argument,		NULL, 'h'},
			{"version",	no_argument,		NULL, 'V'},
			{"copying",	no_argument,		NULL, 'C'},
			{NULL,		0,			NULL,  0 }
		};
		/* *INDENT-ON* */

		c = getopt_long(argc, argv, "qvhVC?", long_options, &option_index);
		if (c == -1)
			break;
		switch (c) {
		case 'q':	/* -q, --quiet */
			output -= output > 0 ? 1 : output;
			break;
		case 'v':	/* -v, --versbose */
			if (optarg == NULL) {
				output += 1;
				break;
			}
			if ((val = strtol(optarg, NULL, 0)) < 0)
				goto bad_option;
			output = val;
			break;
		case 'h':	/* -h, --help */
			help(argc, argv);
			exit(0);
		case 'V':	/* -V, --version */
			version(argc, argv);
			exit(0);
		case 'C':	/* -C, --copying */
			copying(argc, argv);
			exit(0);
		case '?':
		default:
		      bad_option:
			// syntax_error:
			if (optind < argc) {
				fprintf(stderr, "%s: illegal syntax -- ", argv[0]);
				for (; optind < argc; optind++)
					fprintf(stderr, "%s ", argv[optind]);
				fprintf(stderr, "\n");
			}
			usage(argc, argv);
			exit(2);
		}
	}
	exit(0);
}
